src/core/tsi/alts/crypt/aes_gcm.cc

```diff
@@ -35,26 +39,113 @@ constexpr size_t kKdfCounterLen = 6;
 constexpr size_t kKdfCounterOffset = 2;
 constexpr size_t kRekeyAeadKeyLen = kAes128GcmKeyLength;
 
+namespace grpc_core {
+
+GsecKeyFactory::GsecKeyFactory(absl::Span<const uint8_t> key, bool is_rekey)
+    : key_(std::vector<uint8_t>(key.begin(), key.end())), is_rekey_(is_rekey) {}
+
+std::unique_ptr<GsecKeyInterface> GsecKeyFactory::Create() const {
+  return std::make_unique<GsecKey>(absl::MakeConstSpan(key_), is_rekey_);
+}
+
+GsecKey::GsecKey(absl::Span<const uint8_t> key, bool is_rekey)
+    : is_rekey_(is_rekey) {
+  if (is_rekey_) {
+    aead_key_ = std::vector<uint8_t>(kRekeyAeadKeyLen);
+    kdf_buffer_ = std::vector<uint8_t>(EVP_MAX_MD_SIZE);
+    nonce_mask_ = std::vector<uint8_t>(kAesGcmNonceLength);
+    memcpy(nonce_mask_.data(), key.data() + kKdfKeyLen, kAesGcmNonceLength);
+    kdf_counter_ = std::vector<uint8_t>(kKdfCounterLen, 0);
+  }
+  key_ = std::vector<uint8_t>(is_rekey_ ? kKdfKeyLen : key.size());
+  memcpy(key_.data(), key.data(), key_.size());
+}
+
+bool GsecKey::IsRekey() { return is_rekey_; }
+
+absl::Span<const uint8_t> GsecKey::key() { return absl::MakeConstSpan(key_); }
```

Comment:
What's the rationale for using MakeConstSpan for some of these APIs and MakeSpan for others?

Reply:
I think it's fine to leave it as-is: what I was missing before was I did not notice the difference in the return types (i.e. one returns a `uint8_t` span versus the others return a `const uint8_t` span).